Firestore/core/src/firebase/firestore/model/maybe_document.h

```diff
@@ -56,36 +66,30 @@ class MaybeDocument {
     type_ = type;
   }
 
+  virtual bool Equals(const MaybeDocument& other) const;
+
+  friend bool operator==(const MaybeDocument& lhs, const MaybeDocument& rhs);
+
  private:
   Type type_;
   DocumentKey key_;
   SnapshotVersion version_;
 };
 
-/** Compares against another MaybeDocument. */
-inline bool operator<(const MaybeDocument& lhs, const MaybeDocument& rhs) {
-  return lhs.key() < rhs.key();
-}
-
-inline bool operator>(const MaybeDocument& lhs, const MaybeDocument& rhs) {
-  return lhs.key() > rhs.key();
-}
-
-inline bool operator>=(const MaybeDocument& lhs, const MaybeDocument& rhs) {
-  return lhs.key() >= rhs.key();
-}
-
-inline bool operator<=(const MaybeDocument& lhs, const MaybeDocument& rhs) {
-  return lhs.key() <= rhs.key();
+inline bool operator==(const MaybeDocument& lhs, const MaybeDocument& rhs) {
+  return lhs.Equals(rhs);
 }
 
 inline bool operator!=(const MaybeDocument& lhs, const MaybeDocument& rhs) {
-  return lhs.key() != rhs.key() || lhs.type() != rhs.type();
+  return !(lhs == rhs);
 }
 
-inline bool operator==(const MaybeDocument& lhs, const MaybeDocument& rhs) {
-  return lhs.key() == rhs.key() && lhs.type() == rhs.type();
-}
+/** Compares against another MaybeDocument by keys only. */
+struct DocumentKeyComparator {
+  static inline bool Less(const MaybeDocument& lhs, const MaybeDocument& rhs) {
```

Comment:
The interface here should be the same as `std::less<MaybeDocument>`:

```c++
bool operator()(const MaybeDocument& lhs, const MaybeDocument& rhs) const;
```

Note specifically that this means that an instance of the comparator will be required.

Reply:
done